directives/authDirective.js

```diff
@@ -0,0 +1,201 @@
+const { SchemaDirectiveVisitor } = require('apollo-server-express');
+const {
+  UnauthenticatedError,
+  UnauthorizedError,
+  NotFoundError,
+} = require('errors');
+const { defaultFieldResolver } = require('graphql');
+const requestContext = require('talawa-request-context');
+
+const Event = require('../models/Event');
+const Group = require('../models/Group');
+const GroupChat = require('../models/GroupChat');
+const MembershipRequest = require('../models/MembershipRequest');
+const Organization = require('../models/Organization');
+
+class AuthenticationDirective extends SchemaDirectiveVisitor {
+  visitFieldDefinition(field) {
+    const resolver = field.resolve || defaultFieldResolver;
+    field.resolve = async (root, args, context, info) => {
+      const { requires } = this.args;
+
+      if (context.expired)
+        throw new UnauthenticatedError(
+          requestContext.translate('user.notAuthenticated'),
+          'user.notAuthenticated',
+          'userAuthentication'
+        );
+      if (!context.isAuth) {
+        throw new UnauthenticatedError(
+          requestContext.translate('user.notAuthenticated'),
+          'user.notAuthenticated',
+          'userAuthentication'
+        );
+      }
+      if (requires === 'Admin' && args.eventId) {
+        const event = await Event.findOne({ _id: args.eventId });
```

Comment:
Why you are running it at the event?

Reply:
@sumitra19jha In resolvers like `adminRemoveEvent` . We do not have access to the organisation id directly. We have the access to the `eventId`. Therefore, to determine if a user is the admin of the given `Event`. We first have to fetch the `Event` and then take the `organisation id` from there.

Reply:
Our purpose is to classify into these three categories:
1. `Unknown`: If someone is accessing by external service, not through App
2.` User`: If someone is accessing through the app
3. `Admin`: If someone is accessing through the app, who has been given the access of Admin (i.e: Can log into admin portal)

I don't understand what does the event id has to do with it?

Reply:
@sumitra19jha `const isAdmin = org.admins.includes(context.userId);` Sir, in order to determine if a user has admin privileges or not, we need to have the organisation. And, in resolvers like `adminRemoveEvent`, we do not have access to the organisation or the organisation id beforehand. That's why we have to make the calls for event so as to get the corresponding organisation and then determine if the user is admin or not.

In the current iteration of the repo, the same process is being done.

https://github.com/PalisadoesFoundation/talawa-api/blob/develop/resolvers/admin_mutations/admin-remove-event.js#L12-L32

Reply:
Talawa App new version doesn't allow the creation of organisation through the app, I think. This was done to restrict the creation of a large number of organisations through the app. Please go through the documentation once.

Reply:
Can you please explain your solution as in what are you trying to achieve and how?

Reply:
@sumitra19jha So is there any other way to determine admin privileges now?

Reply:
Can you please elaborate on your solution as to what you are trying to achieve and how?
Because it doesn't make sense to classify a user as Admin even in the current version. Suppose if a user has created an organization A and the query is related to organization B, then we shouldn't give the user access to a specific node parameter.

Reply:
> Can you please elaborate on your solution as to what you are trying to achieve and how?
> Because it doesn't make sense to classify a user as Admin even in the current version. Suppose if a user has created an organization A and the query is related to organization B, then we shouldn't give the user access to a specific node parameter.

Sir, this is what I am trying to check in the directive. But, as far as I can think, there is no other way of doing this other than fetching the organisation and checking in the admin array if the current user is in the list or not. To summarise, I am implementing the current method of authentication only, just making it more flexible by separating it from the business logic and doing it on the schema level.

Reply:
Ok!!!

Just do one thing in this PR remove the classification and just implement the separation of authentication from resolver to schema level. 

You told me on slack that you wanted to handle the classification also, so I have given you that suggestion, but we will work on classification afterwards this PR.

Reply:
Update your PR description to fix #297

Reply:
@sumitra19jha Done :+1: